File: src/normal.c
/* normal.c  - core C implementation for Normal distribution
 *
 * To compile with Rtools on Windows:
 *   R CMD SHLIB normal.c
 *
 * This will create normal.dll, which can be loaded from R and
 * called via .C("normal_pdf_c", ...) etc.
 */

#include <math.h>

/* constant */
#ifndef PI
#define PI 3.14159265358979323846
#endif

/* -------------------------
   Internal scalar helpers
   ------------------------- */

/* Normal PDF */
static double normal_pdf(double x, double mu, double sigma) {
    double z = (x - mu) / sigma;
    return exp(-0.5 * z * z) / (sigma * sqrt(2.0 * PI));
}

/* Normal CDF using an approximation to the error function
   Approximation from Abramowitz & Stegun, 7.1.26
*/
static double normal_cdf(double x, double mu, double sigma) {
    double z = (x - mu) / (sigma * sqrt(2.0));
    double t, erf_approx, sign;

    sign = (z < 0.0) ? -1.0 : 1.0;
    z = fabs(z);

    /* erf approximation parameters */
    double a1 = 0.254829592;
    double a2 = -0.284496736;
    double a3 = 1.421413741;
    double a4 = -1.453152027;
    double a5 = 1.061405429;
    double p  = 0.3275911;

    t = 1.0 / (1.0 + p * z);
    erf_approx = 1.0 - (((((a5*t + a4)*t + a3)*t + a2)*t + a1)*t) * exp(-z*z);

    return 0.5 * (1.0 + sign * erf_approx);
}

/* -------------------------
   R .C interface (vectorised)
   All arguments are pointers
   ------------------------- */

/* pdf: compute pdf for vector x of length n
   results written into out[i]
*/
void normal_pdf_c(int *n, double *x,
                  double *mu, double *sigma,
                  double *out) {
    int i;
    for (i = 0; i < *n; i++) {
        out[i] = normal_pdf(x[i], *mu, *sigma);
    }
}

/* cdf: compute cdf for vector x of length n */
void normal_cdf_c(int *n, double *x,
                  double *mu, double *sigma,
                  double *out) {
    int i;
    for (i = 0; i < *n; i++) {
        out[i] = normal_cdf(x[i], *mu, *sigma);
    }
}

/* prob: compute interval probability P(lower[i] <= X <= upper[i])
   lower / upper are vectors of length n
*/
void normal_prob_c(int *n,
                   double *lower, double *upper,
                   double *mu, double *sigma,
                   double *out) {
    int i;
    for (i = 0; i < *n; i++) {
        double cdf_upper = normal_cdf(upper[i], *mu, *sigma);
        double cdf_lower = normal_cdf(lower[i], *mu, *sigma);
        out[i] = cdf_upper - cdf_lower;
    }
}

File: R/normal.R
## normal.R - R wrappers for C-implemented Normal distribution

# 动态加载 C 编译出的 DLL
load_normal_dll <- function(path = NULL) {
  # 默认：在当前工作目录的 src 子目录下找 normal.dll / .so
  if (is.null(path)) {
    path <- file.path(getwd(), "src", paste0("normal", .Platform$dynlib.ext))
  }

  if (!file.exists(path)) {
    stop("Cannot find dynamic library at: ", path)
  }

  # 如果已经加载过，就不用重复加载
  if (!is.loaded("normal_pdf_c")) {
    dyn.load(path)
  }

  invisible(TRUE)
}

# ------------------------------
#   C 实现的 pdf / cdf / prob
# ------------------------------

# PDF 封装（类似于 stats::dnorm）
dnorm_c <- function(x, mean = 0, sd = 1) {
  x <- as.double(x)
  n <- as.integer(length(x))

  res <- .C("normal_pdf_c",
            n     = n,
            x     = x,
            mu    = as.double(mean),
            sigma = as.double(sd),
            out   = double(n))

  res$out
}

# CDF 封装（类似于 stats::pnorm）
pnorm_c <- function(q, mean = 0, sd = 1) {
  q <- as.double(q)
  n <- as.integer(length(q))

  res <- .C("normal_cdf_c",
            n     = n,
            x     = q,
            mu    = as.double(mean),
            sigma = as.double(sd),
            out   = double(n))

  res$out
}

# 区间概率：P(lower <= X <= upper)
pnorm_interval_c <- function(lower, upper,
                             mean = 0, sd = 1) {
  lower <- as.double(lower)
  upper <- as.double(upper)

  if (length(lower) != length(upper)) {
    stop("lower 和 upper 的长度必须相同")
  }

  n <- as.integer(length(lower))

  res <- .C("normal_prob_c",
            n     = n,
            lower = lower,
            upper = upper,
            mu    = as.double(mean),
            sigma = as.double(sd),
            out   = double(n))

  res$out
}
